mantidproject/mantid
# Use RefRoi for conversion to Q in RRO

**Description of work.**

This PR adds a call to `RefRoi` to do the conversion to Q in `ReflectometryReductionOne`. This fixes the workflow algorithm history when summing in lambda because previously this calculation was done within `ReflectometryReductionOne` rather than via a child algorithm.

This PR also includes:
- Up-front validation that `ThetaIn` is provided for the divergent beam case, rather than checking during execution.
- Several unit tests have been added to expand the coverage for cases where the angle is corrected (i.e. by providing `ThetaIn`).
- A typo fix where the minimum wavelength for cropping was not being set correctly for multiple groups.
- A bug fix where spectrum numbers were not being set up correctly for multiple groups.

**Note**
- The output of RefRoi is distribution data and output of ConvertUnits is not distribution. To make things consistent with previous behaviour, for now I've reset the distribution flag to false and the axis label to 'Counts' for the RefRoi case. Unfortunately this defeats the object of this PR slightly because this part of the history is not included. However, this PR is still an improvement so I would like to address the distribution problem in a future PR after discussion with scientists about the best way to do this.

Refs #23012

**To test:**

Should be tested at ISIS with the archive enabled.

- Run the script below.
- Check the history for the `IvsQ` workspaces. They should all use `ConvertUnits` for the conversion to Q except for the two workspaces with `RefRoi` in the name (i.e. when summing in lambda with `ThetaIn` specified).
- Remove `ThetaIn` from the call where the `ReductionType='DivergentBeam'` and it should throw.


```
theta=1.29977  # the angle of specular pixel spectrum 52. Use an ROI centred on this to get the same results whether providing ThetaIn or not.
Load(Filename='INTER00043583.nxs', OutputWorkspace='TOF')

# Sum in wavelength without overriding ThetaIn uses ConvertUnits
ReflectometryReductionOne('TOF', ProcessingInstructions='42-62',
  WavelengthMin=1.5, WavelengthMax=17.0,
  OutputWorkspace='IvsQ_43583_ConvertUnits', SummationType='SumInLambda'
  )
  
# Sum in wavelength and override ThetaIn uses RefRoi
ReflectometryReductionOne('TOF', ProcessingInstructions='42-62',
  WavelengthMin=1.5, WavelengthMax=17.0,
  ThetaIn=theta, OutputWorkspace='IvsQ_RefRoi', SummationType='SumInLambda'
  )

# Using a different value of theta shifts the plot
ReflectometryReductionOne('TOF', ProcessingInstructions='42-62',
  WavelengthMin=1.5, WavelengthMax=17.0,
  ThetaIn=1.1, OutputWorkspace='IvsQ_RefRoi_theta_1.1', SummationType='SumInLambda'
  )
  
# Summing in Q always uses ConvertUnits regardless of whether we provide ThetaIn
ReflectometryReductionOne('TOF', ProcessingInstructions='42-62',
  WavelengthMin=1.5, WavelengthMax=17.0,
  ThetaIn=theta,
  OutputWorkspace='IvsQ_SumInQ_ConvertUnits', SummationType='SumInQ', ReductionType='DivergentBeam'
  )

ReflectometryReductionOne('TOF', ProcessingInstructions='42-62',
  WavelengthMin=1.5, WavelengthMax=17.0,
  OutputWorkspace='IvsQ_SumInQ_bent_ConvertUnits', SummationType='SumInQ', ReductionType='NonFlatSample'
  )
```

![image](https://user-images.githubusercontent.com/12895056/87927305-eb41cb80-ca7a-11ea-9aca-9acd47e9ae50.png)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
